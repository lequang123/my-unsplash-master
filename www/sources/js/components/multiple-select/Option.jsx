import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Option extends Component {
    static propTypes = {
        itemData: PropTypes.object.isRequired,
        id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
        onChange: PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            isExpandedSubData: true
        };
    }

    onChange = event => {
        event.stopPropagation();

        const { itemData, onChange } = this.props;

        const newItemData = {
            key: itemData.key,
            value: itemData.value,
            checked: !itemData.checked
        };
        onChange && onChange(newItemData);
    };

    onClickExpandIcon = () => {
        this.setState(prevState => ({
            isExpandedSubData: !prevState.isExpandedSubData
        }));
    };

    renderSubData = subDataList => {
        return (
            <ul className='sub-data'>
                {
                    this.state.isExpandedSubData &&
                    subDataList.map((item, key) => <li key={`sub-${key}`}>{item}</li>)
                }
            </ul>
        );
    };

    render() {
        const { itemData, id } = this.props;
        const isHasSubData = itemData.subData && itemData.subData.length > 0;
        return (
            <li className="multiple-select-item">
                <input
                    id={id}
                    type="checkbox"
                    className="option-checkbox"
                    checked={itemData.checked}
                    onChange={this.onChange}
                />
                <label className="option-label" htmlFor={id}>
                    {itemData.value}
                </label>
                {
                    isHasSubData &&
                    <span className={`expand-icon fa ${this.state.isExpandedSubData ? 'fa-caret-down' : 'fa-caret-right'}`} onClick={this.onClickExpandIcon} />
                }
                {
                    isHasSubData &&
                    this.renderSubData(itemData.subData)
                }
            </li>
        );
    }
}