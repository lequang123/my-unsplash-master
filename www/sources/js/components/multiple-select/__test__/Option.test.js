import React from 'react';
import Option from '../Option';
import { shallow } from 'enzyme';

describe('Option', () => {
    it('should have correct snapshots', () => {
        // Arrange
        const props = {
            itemData: {
                subData: []
            }
        };

        // Action
        const option = shallow(<Option {...props} />);

        // Assert
        expect(option).toMatchSnapshot();
    });

    it('onChange should run correctly', () => {
        // Arrange
        const itemData = {
            key: 'key',
            value: 'value',
            checked: true,
            subData: []
        };

        const props = {
            itemData,
            onChange: jest.fn()
        };

        const event = { stopPropagation: jest.fn() };
        const option = shallow(<Option {...props} />);

        // Action
        option.instance().onChange(event);

        // Assert
        expect(props.onChange).toHaveBeenCalledWith({
            key: itemData.key,
            value: itemData.value,
            checked: !itemData.checked
        });
    });

    it('onClickExpandIcon change state correctly', () => {
        // Arrange
        const props = {
            itemData : {
                subData: []
            },
            onChange: jest.fn()
        };

        const prevState = { isExpandedSubData: true };
        const option = shallow(<Option {...props} />);
        option.instance().setState(prevState);

        // Action
        option.instance().onClickExpandIcon();

        // Assert
        expect(option.instance().state.isExpandedSubData).toEqual(!prevState.isExpandedSubData);
    });

    it('renderSubData should return expected', () => {
        // Arrange
        const props = {
            itemData : {
                subData: []
            }
        };

        const value = 'value';
        const subDataList = [value];
        const state = { isExpandedSubData: true };
        const expected = '<ul class="sub-data"><li>value</li></ul>';
        
        const option = shallow(<Option {...props} />);
        option.instance().setState(state);

        // Action
        const result = shallow(option.instance().renderSubData(subDataList));

        // Assert
        expect(result.find('ul').props().className).toBe('sub-data');
        expect(result.find('ul').html()).toBe(expected);
    });

    describe('render', () => {
        it('should return regular components', () => {
            // Arrange
            const props = {
                id: 'id',
                itemData : {
                    subData: [],
                    checked: true,
                    value: 'value'
                }
            };

            const onChange = jest.fn();
            const option = shallow(<Option {...props} />);
            option.instance().onChange = onChange;

            // Action
            const result = shallow(option.instance().render());

            // Assert
            expect(result.props().className).toBe('multiple-select-item');
            expect(result.find('input').props().id).toEqual(props.id);
            expect(result.find('input').props().className).toBe('option-checkbox');
            expect(result.find('input').props().checked).toEqual(props.itemData.checked);
            expect(result.find('input').props().onChange).toEqual(onChange);
            expect(result.find('label').props().className).toBe('option-label');
            expect(result.find('label').props().htmlFor).toBe(props.id);
            expect(result.find('label').html()).toContain(props.itemData.value);
        });
        
        it('should return span when having subData', () => {
            // Arrange
            const props = {
                itemData : {
                    subData: [
                        {}
                    ]
                }
            };

            const onClickExpandIcon = jest.fn();
            const renderSubData = jest.fn();
            const state = { isExpandedSubData: true };

            // Action
            const option = shallow(<Option {...props} />);
            option.instance().onClickExpandIcon = onClickExpandIcon;
            option.instance().renderSubData = renderSubData;
            option.instance().setState(state);

            // Assert
            expect(option.find('span').props().className).toBe('expand-icon fa fa-caret-down');
            expect(option.find('span').props().onClick).toEqual(onClickExpandIcon);
            expect(renderSubData).toHaveBeenCalledWith(props.itemData.subData);
        });
    });
});