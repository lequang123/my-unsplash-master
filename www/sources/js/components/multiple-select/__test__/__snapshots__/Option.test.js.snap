// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option should have correct snapshots 1`] = `ShallowWrapper {}`;
